### PR TITLE
Pull command line and config file option readers into separate components

### PR DESCRIPTION
**Pull request by robertsami**
Was getting hard to track how various options effected execution, so pulling reading of options into separate files should hopefully clarify semantics of various knobs going forward.

Also removes a bunch of unused knobs

### CODE REVIEW

File: src/com/oltpbenchmark/CommandLineOptions.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+                "config",
+                true,
+                "Workload configuration file [default: config/workload_all.xml]");
+
+        COMMAND_LINE_OPTS.addOption("v", "verbose", false, "Display Messages");
+        COMMAND_LINE_OPTS.addOption("s", "sample", true, "Sampling window");
+        COMMAND_LINE_OPTS.addOption("im", "interval-monitor", true,
+                "Throughput Monitoring Interval in milliseconds");
+        COMMAND_LINE_OPTS.addOption("ss", false, "Verbose Sampling per Transaction");
+        COMMAND_LINE_OPTS.addOption("ts", "tracescript", true, "Script of transactions to execute");
```
Comment (psudheer21) [inline src/com/oltpbenchmark/CommandLineOptions.java]: we don't use these.
Should we rmove them?

File: src/com/oltpbenchmark/CommandLineOptions.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        COMMAND_LINE_OPTS.addOption(null, "loaderthreads", true, "Number of loader threads (default 10)");
+        COMMAND_LINE_OPTS.addOption(null, "enable-foreign-keys", true, "Whether to enable foregin keys");
+        COMMAND_LINE_OPTS.addOption(null, "create-sql-procedures", true, "Creates the SQL procedures");
+
+        COMMAND_LINE_OPTS.addOption(null, "warmup-time-secs", true, "Warmup time in seconds for the benchmark");
+        COMMAND_LINE_OPTS.addOption(null, "initial-delay-secs", true,
+                "Delay in seconds for starting the benchmark");
+        COMMAND_LINE_OPTS.addOption(null, "num-connections", true, "Number of connections used");
+
+
```
Comment (psudheer21) [inline src/com/oltpbenchmark/CommandLineOptions.java]: Nit: remove these?

File: src/com/oltpbenchmark/DBWorkload.java [hunk 3]
```diff
@@ -252,90 +146,51 @@ public static void main(String[] args) throws Exception {
...
 
-      if (xmlConfig.containsKey("useKeyingTime")) {
-        wrkld.setUseKeyingTime(xmlConfig.getBoolean("useKeyingTime"));
-      }
-      if (xmlConfig.containsKey("useThinkTime")) {
-        wrkld.setUseKeyingTime(xmlConfig.getBoolean("useThinkTime"));
-      }
+      configOptions.getUseKeyingTime().ifPresent(wrkld::setUseKeyingTime);
+      configOptions.getUseThinkTime().ifPresent(wrkld::setUseThinkTime);
+      configOptions.getEnableForeignKeysAfterLoad().ifPresent(wrkld::setEnableForeignKeysAfterLoad);
```
Comment (psudheer21) [inline src/com/oltpbenchmark/DBWorkload.java:176]: This is a cool syntax that i didn't know about..
Awesome.

**Review by psudheer21 — COMMENTED** [review]
Thinking out loud: 
We are anyways using the 2 classes mostly to set the values back in the WorkloadConfiguration.. 
Was wondering if we could just pass the CmdLine Options and the XmlConfiguration directly to the WorkloadConfliguration and then the Get methods can directly access the underlying methods and maybe it will be cleaner?

Let me know what you think.

Regards
Sudheer

**Comment by robertsami** [comment]
> Thinking out loud:
> We are anyways using the 2 classes mostly to set the values back in the WorkloadConfiguration..
> Was wondering if we could just pass the CmdLine Options and the XmlConfiguration directly to the WorkloadConfliguration and then the Get methods can directly access the underlying methods and maybe it will be cleaner?
> 
> Let me know what you think.
> 
> Regards
> Sudheer

I think this is a neat suggestion, but I also like having CommandLineOptions and ConfigFileOptions parsed and read separately, so that all code which references any of their state is clearly documenting where/how said state is configured

at the point where we argue that the code which reads/accesses configuration options should be unified under one component, i might also argue that the configuration should be unified as well, and either all be command line based or configuration file based

lmk if that makes sense to you @psudheer21 or if you still prefer merging them!

File: src/com/oltpbenchmark/CommandLineOptions.java (hunk 1 above)
Comment (robertsami) [reply src/com/oltpbenchmark/CommandLineOptions.java]: we do use interval-monitor, but removing others

**Comment by psudheer21** [comment]
I don't necessarily agree that we have to use just one form of sending arguments.
The things we don't change often like the run_time, the weights etc are part of the xml file and the things we change frequently are part of the command line arguments and I want this to continue in the future as well.

About Having the separate classes for the CommandLine and ConfigFileOptions is fine with me. But can they be part of the Workload Configuration that everyone accesses? I don't like the tens of setters scattered in DBWorkload.java and it would be nice to completely remove them as well.

This can be a follow up review as well.

Regards
Sudheer

**Comment by robertsami** [comment]
> I don't necessarily agree that we have to use just one form of sending arguments.
> The things we don't change often like the run_time, the weights etc are part of the xml file and the things we change frequently are part of the command line arguments and I want this to continue in the future as well.
> 
> About Having the separate classes for the CommandLine and ConfigFileOptions is fine with me. But can they be part of the Workload Configuration that everyone accesses? I don't like the tens of setters scattered in DBWorkload.java and it would be nice to completely remove them as well.
> 
> This can be a follow up review as well.
> 
> Regards
> Sudheer

Oh, you're talking about changing WorkloadConfiguration. Yeah, I think that's a fine plan. Definitely not in the scope of this change though

**Review by psudheer21 — APPROVED** [review]

**Pull request merged**